parcels/uxfieldset.py

```diff
@@ -0,0 +1,189 @@
+        return field[ti,zi,ei]
+    
+    def _node_interp(self, field, time, z, y, x, ei):
+        """Performs barycentric interpolation of a field at a given location."""        
+        ti = 0
+        zi = 0
+        coords =np.deg2rad([[x, y]])
+        n_nodes = self.uxds.uxgrid.n_nodes_per_face[ei].to_numpy()
+        node_ids = self.uxds.uxgrid.face_node_connectivity[ei, 0:n_nodes] 
+        nodes = np.column_stack(
+            (
+                np.deg2rad(self.uxds.uxgrid.node_lon[node_ids].to_numpy()),
+                np.deg2rad(self.uxds.uxgrid.node_lat[node_ids].to_numpy()),
+            )
+        )
+        bcoord = np.asarray(_barycentric_coordinates(nodes, coords))
+        return np.sum(bcoord * field[ti,zi,node_ids].flatten(), axis=0) 
+
+    def get_time_range(self):
+        return self.uxds.time.min().to_numpy(), self.uxds.time.max().to_numpy()
+    
+    def _point_is_in_face(self, y, x, ei):
+        "Checks if a point is inside a given face id "
+        #ti, zi, fi = self.unravel_index(particle.ei) # Get the time, z, and face index of the particle
+        fi = ei
+
+        # Check if particle is in the same face, otherwise search again.
+        n_nodes = self.uxds.uxgrid.n_nodes_per_face[fi].to_numpy()
+        node_ids = self.uxds.uxgrid.face_node_connectivity[fi, 0:n_nodes]
+        nodes = np.column_stack(
+            (
+                np.deg2rad(self.uxds.uxgrid.node_lon[node_ids].to_numpy()),
+                np.deg2rad(self.uxds.uxgrid.node_lat[node_ids].to_numpy()),
+            )
+        )
+
+        coord = np.deg2rad([x, y])
+        bcoord = np.asarray(_barycentric_coordinates(nodes, coord))
+        if ( not (bcoord >= 0).all() ) and (not (bcoord <= 1.0).all()):
+            return False
+        
+        return True
+
+    def eval(self, field_names, time, z, y, x, ei: int=None, applyConversion=True):
+        
+        res = {}
+
+        if ei is not None:
+            fi = ei
+            if not self._point_is_in_face(y,x,ei):
+                # If the point is not in the previously defined face, then
+                # search for the face again.
+                # To do : Update the search here to do nearest neighbors search, rather than spatial hash - joe@fluidnumerics.com
+                print(f"Position : {x}, {y}")
+                print(f"Hash indices : {self._spatialhash._hash_index2d(np.deg2rad([[x,y]]))}")
+                fi, bcoords = self._spatialhash.query([[x,y]]) # Get the face id for the particle  
+                fi = fi[0] 
+                print(f"Face index (updated): {fi}")
+                print(f"Barycentric coordinates (updated): {bcoords}")
+
+        for f in field_names:
+            field = getattr(self.uxds, f)
+            face_registered = ("n_face" in field.dims)
+
+            if face_registered:
+                r = self._face_interp(field, time, z, y, x, fi)
+            else:
+                r = self._node_interp(field, time, z, y, x, fi)
+
```

Comment:
For 3-D interpolation, we need to add vertical interpolation. Currently, all interpolation uses the fields at the upper most vertical layer